### DISCUSSION TITLE
How to get webscoket context in `onClose`

### DISCUSSION THREAD

**Discussion by alucardu** [enisdenjo/graphql-ws#462] [Q&A] [answered]
I want to notify users (and clean up my database) when a user disconnects from the websocket:
```javascript
const serverCleanup = useServer(
  {
    schema,
    onConnect: async () => {
      console.log('User connected')
    },
    onClose() {
      console.log('User disconnected')
      userResolvers.Mutation.deleteUser({
        name: 'John'
      })
    },
  },
  wsServer
)
```
In my server log I do see `User disconnected { name: 'jan' }` but I of course need the user id to correctly remove the user from the database. I'm using Apollo client and server as my graphql solution. 

I do have a websocket connection running, and I'm also happy with it:
![image](https://user-images.githubusercontent.com/779921/223740538-785fd7c0-a7e7-4b40-b459-836bf81abc16.png)

So what I would like to know is how do I get (all) websocket information in the `onClose()` function so I can delete the user that has closed the connection.

**Comment by enisdenjo** [comment, marked as answer]
You can use the [context extra field](https://github.com/enisdenjo/graphql-ws/blob/7422ede00985eaa20137bd2650e6e9d19992fcff/docs/interfaces/server.Context.md#extra) which is shared between the hooks for distinct connections:

```diff
const serverCleanup = useServer(
  {
    schema,
-   onConnect: async () => {
+   onConnect: async (ctx) => {
+     ctx.extra.userName = getUserName()
      console.log('User connected')
    },
-   onClose() {
+   onClose(ctx)
      console.log('User disconnected')
      userResolvers.Mutation.deleteUser({
-       name: 'John'
+       name: ctx.extra.userName
      })
    },
  },
  wsServer
)
```

**Reply by alucardu** [reply]
Thank you for the suggestion @enisdenjo. In my app users can create a username and connect that to a session. So initially there's no information available.

Should I then add the created user to the websocket and extract it like so:
```typescript
private subscribeToUsers(user: User): void {
  this.apollo.subscribe({
    query: USER_CREATED,
    variables: {
      ...user
    },
  }).subscribe(({
    next: (data) => {
      console.log(data);
    }
  }));
}
```` 

```javascript
const serverCleanup = useServer(
  {
    schema,
    onConnect: async () => {
      console.log('User connected')
    },
    onClose(ctx) {
      console.log('onClose', ctx.extra.userId)
    },
    context(ctx, msg, args) {
      if (msg?.payload?.operationName === 'userCreated') {
        ctx.extra.userId = msg.payload.variables.id
      }
    }
  },
  wsServer
)
```
When I create a new user I subscribe to it, then I can extract it from the `msg` argument in the `context()` and use that value to put it in the extra context. Seems to work.

**Reply by enisdenjo** [reply]
This behaviour will overwrite the current use in the context for each `userCreated` operation, can be quite unsafe. If you can guarantee that `userCreated` will be ran just once - then fine; otherwise I'd rethink this if I were you.
